### PR TITLE
PWX-25748: Add grafana spec support for PX dashboards

### PR DESCRIPTION
**Pull request by ggriffiths**
**What this PR does / why we need it**:
Add grafana support in the operator spec

**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:
Tested with port forward that grafana UI can be viewed and dashboards are all working.

### CODE REVIEW

File: drivers/storage/portworx/component/monitoring.go [hunk 1]
```diff
@@ -1,8 +1,11 @@
 package component
 
 import (
+	"context"
```
Comment (piyush-nimbalkar) [inline drivers/storage/portworx/component/monitoring.go:4]: Let's extract the Grafana components into a different component called grafana.
The main reason being the trigger/flag for grafana is different from the prometheus stack. In the current state, the `IsEnabled` method will allow grafana installation even if just prometheus is enabled.

File: drivers/storage/portworx/component/monitoring.go [hunk 2]
```diff
@@ -99,6 +157,10 @@ func (c *monitoring) Delete(cluster *corev1.StorageCluster) error {
 	if err != nil && !metaerrors.IsNoMatchError(err) {
 		return err
 	}
+	err = k8sutil.DeleteDeployment(c.k8sClient, GrafanaDeploymentName, cluster.Namespace, *ownerRef)
+	if err != nil && !metaerrors.IsNoMatchError(err) {
```
Comment (piyush-nimbalkar) [inline drivers/storage/portworx/component/monitoring.go:161]: nit: Don't have to check for the `metaerrors.IsNoMatchError` error as we are just deleting a deployment and not a custom resource.

File: drivers/storage/portworx/component/monitoring.go [hunk 3]
```diff
@@ -159,6 +221,185 @@ func (c *monitoring) createPrometheusRule(
...
+				)
+			}
+		}
+	}
+
+	// If resources is specified in the spec, the resources specified by annotation (such as portworx.io/autopilot-cpu)
+	// will be overwritten.
+	if cluster.Spec.Autopilot.Resources != nil {
+		deployment.Spec.Template.Spec.Containers[0].Resources = *cluster.Spec.Autopilot.Resources
+	}
```
Comment (piyush-nimbalkar) [inline drivers/storage/portworx/component/monitoring.go:398]: Let's remove this as it does not apply here.

File: drivers/storage/portworx/component/monitoring.go [hunk 4]
```diff
@@ -31,14 +39,61 @@ const (
...
+		{
+			Name:      "dashboard-templates",
+			MountPath: "/var/lib/grafana/dashboards",
+			VolumeSource: v1.VolumeSource{
+				ConfigMap: &v1.ConfigMapVolumeSource{
+					LocalObjectReference: v1.LocalObjectReference{
+						Name: "grafana-dashboard-config",
+					},
+				},
+			},
```
Comment (piyush-nimbalkar) [inline drivers/storage/portworx/component/monitoring.go:87]: Aren't these the same volumes with just different mount points? Is this how it is supposed to be?

File: drivers/storage/portworx/component/monitoring.go [hunk 5]
```diff
@@ -31,14 +39,61 @@ const (
...
 
+var (
+	grafanaDeploymentVolumes = []corev1.VolumeSpec{
+		{
+			Name:      "grafana-source-config",
+			MountPath: "/etc/grafana/provisioning/datasources",
+			VolumeSource: v1.VolumeSource{
+				ConfigMap: &v1.ConfigMapVolumeSource{
+					LocalObjectReference: v1.LocalObjectReference{
+						Name: "grafana-dashboards",
```
Comment (piyush-nimbalkar) [inline drivers/storage/portworx/component/monitoring.go:62]: We will have to create these config maps, right?

File: drivers/storage/portworx/component/monitoring.go [hunk 6]
```diff
@@ -31,14 +39,61 @@ const (
...
+	grafanaPort          = 3000
+	grafanaLoginEndpoint = "/login"
 	pxPrometheusRuleFile = "portworx-prometheus-rule.yaml"
 )
 
+var (
+	grafanaDeploymentVolumes = []corev1.VolumeSpec{
+		{
+			Name:      "grafana-source-config",
+			MountPath: "/etc/grafana/provisioning/datasources",
```
Comment (piyush-nimbalkar) [inline drivers/storage/portworx/component/monitoring.go:58]: When configuring the datasource, we will have to add our operator deployed prometheus to it. We should probably only create the dashboard configmap or even grafana only if our prometheus is deployed. Wdyt?

File: drivers/storage/portworx/component/monitoring.go [hunk 7]
```diff
@@ -195,6 +436,13 @@ func (c *monitoring) retryCreate(
 	return true
 }
 
+func (c *monitoring) getDesiredGrafanaImage(cluster *corev1.StorageCluster) string {
```
Comment (piyush-nimbalkar) [inline drivers/storage/portworx/component/monitoring.go:439]: We will have to make changes to [manifest.go](https://github.com/libopenstorage/operator/blob/master/drivers/storage/portworx/manifest/manifest.go) and [portworx.go](https://github.com/libopenstorage/operator/blob/master/drivers/storage/portworx/portworx.go#L560) to get the grafana image from version manifest and then set it in the `status.desiredImages`.

**Review by piyush-nimbalkar — COMMENTED** [review]
Generally looks good. Let's add UTs after the comments are addressed.

File: drivers/storage/portworx/component/monitoring.go (hunk 2 above)
Comment (ggriffiths) [reply drivers/storage/portworx/component/monitoring.go]: Removed this

File: drivers/storage/portworx/component/monitoring.go (hunk 3 above)
Comment (ggriffiths) [reply drivers/storage/portworx/component/monitoring.go]: Removed

File: drivers/storage/portworx/component/monitoring.go (hunk 4 above)
Comment (ggriffiths) [reply drivers/storage/portworx/component/monitoring.go]: Ah, this was on an older commit. Fixed this.

File: drivers/storage/portworx/component/monitoring.go (hunk 5 above)
Comment (ggriffiths) [reply drivers/storage/portworx/component/monitoring.go]: Yes creating them now in grafana.go

File: deploy/grafana/grafana-datasource.yaml [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+      # <bool> mark as default datasource. Max one per org
+      isDefault:
+      # <map> fields that will be converted to json and stored in json_data
+      jsonData:
+        graphiteVersion: "1.1"
+        tlsAuth: false
+        tlsAuthWithCACert: false
+      # <string> json object of data that will be encrypted.
+      secureJsonData:
+        tlsCACert: "..."
```
Comment (Pure-AdamuKaapan) [inline deploy/grafana/grafana-datasource.yaml:43]: Does this get replaced by something else later on?

File: drivers/storage/portworx/component/grafana.go [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+	}
+	err = k8sutil.DeleteConfigMap(c.k8sClient, PXGrafanaDatasourceConfigConfigMap, cluster.Namespace, *ownerRef)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *grafana) MarkDeleted() {}
```
Comment (Pure-AdamuKaapan) [inline drivers/storage/portworx/component/grafana.go:177]: Is this intentionally empty? If so, can we add a comment as to why this is okay/what it's "not doing"?

File: drivers/storage/portworx/testspec/grafanaDeployment.yaml [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+          name: px-grafana-dashboard-config
+        name: grafana-dashboard-config
+      - configMap:
+          defaultMode: 420
+          name: px-grafana-dashboards-json
+        name: grafana-dashboard-templates
+      - configMap:
+          defaultMode: 420
+          name: px-grafana-datasource-config
+        name: grafana-source-config
```
Comment (Pure-AdamuKaapan) [inline drivers/storage/portworx/testspec/grafanaDeployment.yaml:77]: [nit] ensure newlines at end of files

File: drivers/storage/portworx/testspec/grafanaService.yaml [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+  name: px-grafana
+  namespace: kube-test
+  labels:
+    app: grafana
+spec:
+  ports:
+    - port: 3000
+  selector:
+    app: grafana
+  type: ClusterIP
```
Comment (Pure-AdamuKaapan) [inline drivers/storage/portworx/testspec/grafanaService.yaml:13]: [nit] newlines at end of files

**Review by Pure-AdamuKaapan — APPROVED** [review]
Overall LGTM though I had a couple minor comments

File: deploy/grafana/grafana-datasource.yaml (hunk 8 above)
Comment (ggriffiths) [reply deploy/grafana/grafana-datasource.yaml:43]: I think this was left as a template for customers to add cert info. I'm removed it for this

File: drivers/storage/portworx/component/grafana.go (hunk 9 above)
Comment (ggriffiths) [reply drivers/storage/portworx/component/grafana.go:177]: Yeah, most components have this as empty. I've added a comment

**Pull request merged**